confgenerator/resourcedetector/detector.go

```diff
@@ -0,0 +1,45 @@
+package resourcedetector
+
+import (
+	"regexp"
+
+	gcp_metadata "cloud.google.com/go/compute/metadata"
+)
+
+type Detector interface {
+	GetMetaLabels() (Metalabels, error)
+}
+
+// Get a Detector for the current environment, which can be used
+// to get all the metadata
+func GetDetector() Detector {
+	switch {
+	case gcp_metadata.OnGCE():
+		return NewGCEDetector(nil)
+	default:
+		return EmptyDetector{}
+	}
+}
+
+type LabelName string
+
+type LabelValue string
+
+type Metalabels map[LabelName]LabelValue
```

Comment:
The contract we expose should probably be more clear - instead of having `map[string]string` it could be a struct that has fields for all the attributes. This way it is part of the public interface what fields are available as part of the detector, instead of maintaining another set of key labels.